lib/Dialect/TritonGPU/Transforms/Prefetch.cpp

```diff
@@ -291,6 +291,16 @@ scf::ForOp Prefetcher::createNewForOp() {
   mapping.map(forOp.getInductionVar(), newForOp.getInductionVar());
 
   for (Operation &op : forOp.getBody()->without_terminator()) {
+    // If we find anything that depends on a dot op, stop inserting before the
+    // dot, and add to the end of the block instead.
+    for (auto operand : op.getOperands()) {
+      if (auto def = operand.getDefiningOp()) {
+        auto dot = dyn_cast<triton::DotOp>(def);
+        if (dot && dots.contains(dot)) {
+          builder.setInsertionPointToEnd(newForOp.getBody());
+        }
+      }
+    }
```

Comment:
ops could have nested regions with use of tt.dot?

something like:
```
%d = tt.dot
scf.if 
   %user = user_op %d
```

Reply:
Good catch. I guess we can stop when hit a region op. So at end of each dot op, we set the insertion point to the last decomposed small dot. When going through the later ops in the loop, we can advance the insertion point to the end of the new forOp when needed.

Reply:
Yeah, the current implementation may not work well if the dot is in a if-block or there are other control flows. Maybe just bail out when there is any control flow in the loop body?

Also IIUC, the code following the original dot is cut in half by the dot dependency op. The first half will be placed before the last dot. This seems quite random and I'm not sure it's always beneficial. For example,  there could be expansive operations that originally don't block the dot but now do, such as a async_wait for unrelated loads for other operations. Correct me if this is not valid.

Reply:
🤦 good catch, I forgot entirely about control flow.  I've updated the PR to stop sinking the dot if any control flow is encountered.  (I guess I could recurse into the control flow region to see if the dot is used inside... but I'm not sure it's worth trying to optimize that case)